### ISSUE TITLE
Tri-tri intersect returns unreasonable intersection segment when tris have shared edge.

### ISSUE THREAD

**Issue by ramcdona** [erich666/jgt-code#10]
I've run into an issue with:

Volume_08/Number_1/Guigue2003/tri_tri_intersect.c

See the Matlab code below that will create this figure:

![tri-tri-error](https://github.com/erich666/jgt-code/assets/1329600/5589d1c6-98f3-4157-94ef-95b088793ac1)

```
t0 = [-2.000000000000000000000000e+02 -7.661661997582373651027865e+00 1.107006972331142975463081e+02; % A
      -2.000000000000000000000000e+02 -1.532332399516475796019677e+01 1.214013944662285808817614e+02; % B
      -2.000000000000000000000000e+02 -1.041705198583709623960658e+01 1.221265882878931563482183e+02; % C
      -2.000000000000000000000000e+02 -7.661661997582373651027865e+00 1.107006972331142975463081e+02];

t1 = [-2.000000000000000000000000e+02 -1.532332399516475796019677e+01 1.214013944662285808817614e+02; % D
      -2.000147534839701108921872e+02 -5.208525992918542790732772e+00 1.110632941439465781741092e+02; % E
      -2.000000000000000000000000e+02 -1.041705198583709623960658e+01 1.221265882878931563482183e+02; % F
      -2.000000000000000000000000e+02 -1.532332399516475796019677e+01 1.214013944662285808817614e+02];

e0 = [-2.000000000000000000000000e+02 -1.148161784440646755456328e+02 1.066954011562485078457030e+02; %G
      -2.000000000000000000000000e+02 -1.041705198583709623960658e+01 1.221265882878931563482183e+02]; % H
figure( 2 );
plot3( t0(:,1), t0(:,2), t0(:,3) )
hold on;
plot3( t1(:,1), t1(:,2), t1(:,3) )
plot3( e0(:,1), e0(:,2), e0(:,3), '-o' )
hold off;
axis equal;
axis off;
```

For these two triangles, points B and D are congruent and points C and F are congruent.  I.e. they share an edge.

In this case, one of the intersection end points comes back as one of the endpoints of the shared edge.  I would expect the other intersection end point to come back as the other endpoint of the shared edge.  However, instead it is a point located far away.

Point H is at point C/F.
Point G is far away.

All intersection points should lie on/within both triangles.

**Comment by erich666** [comment]
Sorry, just noticed this (I finally fixed my GIthub notifications so they work correctly). I've written the first author and hope for a response.

**Comment by pguigue** [comment]
Hi,
This problem results from a rounding error issue similar to the one detected in https://github.com/erich666/jgt-code/issues/9 ,
that is,  in some edge cases and because of the limited precision of IEEE floating point arithmetic, some coplanar triangles' vertices are not detected as they should be.

To use the proposed fix (https://github.com/erich666/jgt-code/pull/12) you just have to set the USE_EPSILON_TEST macro to true in the code (it is set to false by default).
The new EPSILON value is set to 1e-16 in the PR to cover the case above, the greater computed faulty determinant being evaluated to about 6.93e-18.

**Comment by erich666** [comment]
Hopefully this fix (merged) addresses this bug. Let us know, @ramcdona, if it doesn't.

**Comment by ramcdona** [comment]
Thanks for the patch, I'll give it a shot.

I had moved away from Thomas Möller's version exactly because of this epsilon test and the hope that this code would be more robust to floating point issues.

I've gone a ways down the rabbit hole of tri-tri intersection codes in the literature.  I've found a few that claim to be provably robust (using robust geometric predicates for all conditionals).  Unfortunately, they don't seem to have a publicly released implementation.

It would seem that despite this being a 'solved problem', there is perhaps need for yet another solution.
